keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,449 @@
+
+API fields in the Workload, LocalQueue and ClusterQueue objects control
+different preemption policies.
+
+## Motivation
+
+when ClusterQueues under their minimum quota lend resources, they should
+be able to recover those resources fast, to be able to admit workloads
+when there are sudden spikes.
+
+Currently, the only mechanism to recover those resources is to wait for
+Workloads to finish, which is generally unbounded.
+
+### Goals
+
+- Preempt Workloads from ClusterQueues borrowing resources when other
+  ClusterQueues in the cohort, under their minimum quota, need the resources.
+- Introduce API fields in Workload, LocalQueue and ClusterQueue to inform
+  which Workloads can be preempted and in which order.
+
+### Non-Goals
+
+- Graceful termination of Workloads is left to the controller that integrates
+  with the particular workload API.
+  For example, the integration with Job uses the [suspend field](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job).
+- Partial workload suspension is not supported.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This enhancement proposes the introduction of the following APIs to influence
+Workload preemption:
+- A field in the Workload to describe the cost of preempting a Workload at a
+  given time.
+- A field in the LocalQueue to set Workload defaults.
+- A field in the ClusterQueue to influence how long to wait before starting
+  preemption.
+
+And an algorithm for selecting a set of Workloads using borrowed quota to be
+preempted when a Workload in a ClusterQueue is under the min quota.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a batch user, I know how safe it is to disrupt some workloads after they
+have been running for some time.
+
+An sample Workload has the following characteristics:
+Early in their execution, the Workload hasn't made much progress, so it can be
+preempted. Later, the Worload is on the path of doing significant progress, so
+it's best not to disturb it. Lastly, the Workload is expected to have made
+some checkpoints, so it's ok to disturb it.
+
+This can be expressed with the following configuration:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: Workload
+metadata:
+  name: my-workload
+spec:
+  preemption:
+    disruptionCostMilestones:
+    - seconds: 60
+      cost: 100
+    - seconds: 600
+      cost: 0
+```
```

Comment:
At first glance this doesn't look simple. Did we base this API or the general semantics on related batch schedulers? How can we assess or have confidence if this is this what users want or even able to express? 

Also, it is possible that cost is not associated with the run time, should we consider a dynamic approach to setting the cost?

Reply:
I learned this acronym/principle from Clayton the other day, YAGNI :)

I am concerned that we are adding an API that will not be used, so I suggest we start simple and drop this API. We fix the ranking by priority then runtime (based on buckets like we did for replicasets downscaling). We can always add the API later if someone comes forward with a solid use case.

Reply:
I guess we can say that the request in https://github.com/kubernetes-sigs/kueue/issues/83#issuecomment-1224602577 is mostly covered by the second story and drop this for later.

Reply:
Moved to Alternatives